text/0000-stabilize-catch-panic.md

```diff
@@ -0,0 +1,345 @@
+rare, **when it happens it is not a programmer error**. In short, panics are
+roughly analogous to an opaque "an unexpected error has occurred" message.
+
+Stabilizing `catch_panic` does little to change the tradeoffs around `Result`
+and `panic` that led to these conventions.
+
+## Why remove the bounds?
+
+There are a few reasons to remove the `'static` and `Send` bounds on the
+`catch_panic` function:
+
+* One of the primary use cases of `catch_panic` is in an FFI context, where lots
+  of `*mut` and `*const` pointers are flying around. These two types aren't
+  `Send` by default, so having their values cross the `catch_panic` boundary
+  would be highly un-ergonomic (albeit still possible). As a result, this RFC
+  proposes removing the `Send` bound from the function.
+
+* A reason to remove the `'static` bound is that it doesn't provide rock-solid
+  exception-safety mitigation. Using thread-local storage it's possible to
+  share mutable data across a call to `catch_panic` even if that data isn't
+  `'static`.
+
+* Borrowed data, in particular, is helpful for thread pools that need
+  to execute closures with borrowed data within them; essentially, the worker
+  threads are executing multiple "semantic threads" over their lifetime, and the
+  `catch_panic` boundary represents the end of these "semantic threads".
+
+# Drawbacks
+
+A drawback of this RFC is that it can water down Rust's error handling story.
+With the addition of a "catch" construct for exceptions, it may be unclear to
+library authors whether to use panics or `Result` for their error types. As we
+discussed above, however, Rust's design around error handling has always had to
+deal with these two strategies, and our conventions don't materially change by
+stabilizing `catch_panic`.
+
+# Alternatives
+
+One alternative, which is somewhat more of an addition, is to have the standard
+library entirely abandon all exception safety mitigation tactics. As explained
+in the motivation section, exception safety will not lead to memory unsafety
+unless paired with unsafe code, so it is perhaps within the realm of possibility
+to remove the tactics of poisoning from mutexes and simply require that
+consumers deal with exception safety 100% of the time.
+
+This alternative is often motivated by saying that there are enough methods to
+subvert the default mitigation tactics that it's not worth trying to plug some
+holes and not others. Upon closer inspection, however, the areas where safe code
+needs to worry about exception safety are isolated to the single-threaded
+situations. For example `RefCell`, destructors, and `catch_panic` all only
+expose data possibly broken through a panic in a single thread.
+
+Once a thread boundary is crossed, the only current way to share data mutably is
+via `Mutex` or `RwLock`, both of which are poisoned by default. This sort of
+sharing is fundamental to threaded code, and poisoning by default allows safe
```

Comment:
That's not strictly true, there are also atomics. I doubt this will be a problem in many cases, but if you e g implement a spinlock using atomics, you'll have to consider poisoning the spinlock as well.